File: Algorithms/p238_Product_of_Array_Except_Self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        nums_len = len(nums)
        res = [None] * nums_len
        res[0] = 1
        for i in range(1,nums_len):
            res[i] = res[i-1] * nums[i-1]
        r_product = nums[-1]
        for i in range(nums_len-2,-1,-1):
            res[i] *= r_product
            r_product *= nums[i]
        return res

File: Algorithms/p238_Product_of_Array_Except_Self/p238_Product_of_Array_Except_Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int len = nums.size();
        vector<int> res(nums);
        res[0] = 1;
        for(int i=1;i<len;i++){
            res[i] = res[i-1] * nums[i-1];
        }
        int r_product = nums[len-1];
        for(int i=len-2;i>=0;i--){
            res[i] *= r_product;
            r_product *= nums[i];
        }
        return res;
    }
};
